File: MessageList/app/src/main/java/ru/startandroid/messagelist/app/dagger/module/AppSubComponentsModule.java
package ru.startandroid.messagelist.app.dagger.module;

import dagger.Module;
import dagger.Provides;
import dagger.multibindings.ClassKey;
import dagger.multibindings.IntoMap;
import ru.startandroid.messagelist.base.dagger.BaseComponentBuilder;
import ru.startandroid.messagelist.ui.message.list.MessageListFragment;
import ru.startandroid.messagelist.ui.message.list.dagger.MessageListComponent;

@Module(subcomponents = {MessageListComponent.class})
public class AppSubComponentsModule {

    @Provides
    @IntoMap
    @ClassKey(MessageListFragment.class)
    BaseComponentBuilder provideMessageListComponentBuilder(MessageListComponent.Builder builder) {
        return builder;
    }

}

File: Vocabulary/app/src/main/java/ru/startandroid/vocabulary/app/dagger/AppSubComponentsModule.kt
package ru.startandroid.vocabulary.app.dagger

import dagger.Module
import dagger.Provides
import dagger.multibindings.ClassKey
import dagger.multibindings.IntoMap
import ru.startandroid.vocabulary.base.dagger.BaseComponent
import ru.startandroid.vocabulary.base.dagger.BaseModule
import ru.startandroid.vocabulary.base.dagger.ComponentBuilder
import ru.startandroid.vocabulary.ui.dictionary.details.dagger.WordDetailsDagger
import ru.startandroid.vocabulary.ui.dictionary.details.mvp.WordDetailsFragment
import ru.startandroid.vocabulary.ui.dictionary.list.dagger.WordListDagger
import ru.startandroid.vocabulary.ui.dictionary.list.mvp.WordListFragment

@Module(subcomponents = arrayOf(WordListDagger.WordListComponent::class, WordDetailsDagger.WordDetailsComponent::class))
class AppSubComponentsModule {

    @Provides
    @IntoMap
    @ClassKey(WordListFragment::class)
    fun provideWordListFragmentBuilder(builder: WordListDagger.WordListComponent.Builder): ComponentBuilder<out BaseComponent<*>, out BaseModule> {
        return builder
    }


    @Provides
    @IntoMap
    @ClassKey(WordDetailsFragment::class)
    fun provideWordDetailsFragmentBuilder(builder: WordDetailsDagger.WordDetailsComponent.Builder): ComponentBuilder<out BaseComponent<*>, out BaseModule> {
        return builder
    }

}